server/libs/infra/src/search/typesense.repository.ts

```diff
@@ -0,0 +1,181 @@
+import { DocumentSchema, SearchResponse } from 'typesense/lib/Typesense/Documents';
+import { AlbumEntity, AssetEntity } from '../db';
+import { albumSchema } from './schemas/album.schema';
+import { assetSchema } from './schemas/asset.schema';
+
+const schemaMap: Record<SearchCollection, CollectionCreateSchema> = {
+  [SearchCollection.ASSETS]: assetSchema,
+  [SearchCollection.ALBUMS]: albumSchema,
+};
+
+const schemas = Object.entries(schemaMap) as [SearchCollection, CollectionCreateSchema][];
+
+// TODO: look at using an external file for advanced configuration
+export const typesense = new Client({
+  nodes: [
+    {
+      host: process.env.TYPESENSE_HOST || 'typesense',
+      port: Number(process.env.TYPESENSE_PORT) || 8108,
+      protocol: process.env.TYPESENSE_PROTOCOL || 'http',
+    },
+  ],
+  apiKey: process.env.TYPESENSE_API_KEY || 'should-not-be-used',
+  numRetries: 3,
+  connectionTimeoutSeconds: 10,
+});
+
+@Injectable()
+export class TypesenseRepository implements ISearchRepository {
+  private client = typesense;
+  private logger = new Logger(TypesenseRepository.name);
+  private queue: Record<SearchCollection, any[]> = {
+    [SearchCollection.ASSETS]: [],
+    [SearchCollection.ALBUMS]: [],
+  };
+
+  constructor() {
+    setInterval(() => this.flush(), 5_000);
+  }
+
+  async setup(): Promise<void> {
+    // upsert collections
+    for (const [collectionName, schema] of schemas) {
+      const collection = await this.client
+        .collections(schema.name)
+        .retrieve()
+        .catch(() => null);
+      if (!collection) {
+        this.logger.log(`Creating schema: ${collectionName}/${schema.name}`);
+        await this.client.collections().create(schema);
+      } else {
+        this.logger.log(`Schema up to date: ${collectionName}/${schema.name}`);
+      }
+    }
+  }
+
+  async checkMigrationStatus(): Promise<SearchCollectionIndexStatus> {
+    const migrationMap: SearchCollectionIndexStatus = {
+      [SearchCollection.ASSETS]: false,
+      [SearchCollection.ALBUMS]: false,
+    };
+
+    // check if alias is using the current schema
+    const { aliases } = await this.client.aliases().retrieve();
+    this.logger.log(`Alias mapping: ${JSON.stringify(aliases)}`);
+
+    for (const [aliasName, schema] of schemas) {
+      const match = aliases.find((alias) => alias.name === aliasName);
+      if (!match || match.collection_name !== schema.name) {
+        migrationMap[aliasName] = true;
+      }
+    }
+
+    this.logger.log(`Collections needing migration: ${JSON.stringify(migrationMap)}`);
+
+    return migrationMap;
+  }
+
+  async index(collection: SearchCollection, item: AssetEntity | AlbumEntity, immediate?: boolean): Promise<void> {
+    const schema = schemaMap[collection];
+
+    if (immediate) {
+      await this.client.collections(schema.name).documents().upsert(item);
+      return;
+    }
+
+    this.queue[collection].push(item);
```

Comment:
What's the reason for the asynchronous queue stuff? Just request batching?

Reply:
Yeah exactly. Typesense docs mentioned it is faster/more efficient this way.